### ISSUE TITLE
[Feature]: Stop Training from wandb UI/API

### ISSUE THREAD

**Issue by dreamflasher** [wandb/client#3366]
### Description

We'd like to be able to stop a training run from wandb (ideally from the API).

### Suggested Solution

Add a button to the web UI "stop run" and add a flag to each run "stop requested", expose via API to set this flag. Every time a wandb function is called for that run (ie. training logging) it checks if "stop requested" is true, and if yes terminates the current process (or raise an exception that can be handled by the training code, ideally make it an option if it raises an exception or just terminates that code).

**Comment by nate-wandb** [comment]
Hi @dreamflasher,
Thank you for the feature request! Stopping the run via the UI can be done by going to the run overview tab in the UI and clicking the stop icon next to the run state:
![Screen Shot 2022-03-14 at 10 10 35 AM](https://user-images.githubusercontent.com/97066933/158215583-0dbddce1-517d-40b6-969b-69a264181f6e.png)
Let me know if this differs from what you were looking for.

The rest of your feature request does not currently exist so I will pass this onto our engineering team for you and let you know if we make any progress towards this.

Thank you,
Nate

**Comment by dreamflasher** [comment]
That's exactly what I have been looking for – in the API. Do you expose this feature in the API?

**Comment by nate-wandb** [comment]
No, unfortunately stopping a run via the API isn't currently possible so I'll put in a feature request for this. The way you described the stopping process is essentially how the stop button on the UI works so I imagine that if we added support to do this through the API it would work the same way.

**Comment by dreamflasher** [comment]
Thank you very much for adding a feature request for it. Yes, it sounds exactly like what I had in mind.

**Comment by nate-wandb** [comment]
You're welcome! I'll let you know if we have any progress on this. Thank you!

**Comment by vict0rsch** [comment]
Hey! Any news on the stop-from-api front?

**Comment by nate-wandb** [comment]
Hi @vict0rsch, this has been put on backlog but I can bump up the priority since it has come up again to see if this is something we can get implemented.

**Comment by davidgilbertson** [comment]
I don't see this button, under what circumstances will it be visible?
![image](https://user-images.githubusercontent.com/4443482/211717268-b3a1b136-0980-4d5d-9e08-8452eb8312b3.png)

**Comment by vict0rsch** [comment]
It's really important for large sweeps, for handling multiple projects and users as well as for the integration with other software that provide a Python interface

**Comment by nate-wandb** [comment]
@vict0rsch thank you for the context and I've passed this along to the team as well. I'll send an update once this has gone through triage again. 

@davidgilbertson is that run part of a team? You can only "edit" runs that are ran to your personal entity or you are the admin for the team. If you don't think this is the issue, could you possibly send me a link to the run?

**Comment by davidgilbertson** [comment]
This is a 'personal' (free) account, and I've just started logging runs, I haven't set up anything (teams, permissions, etc) that wasn't required to get started.

I can tell you the URL but it isn't public: https://wandb.ai/gilbertson-david/adaptive-sampler/runs/0tl7x3wu/overview?workspace=user-gilbertson-david

**Comment by nate-wandb** [comment]
@davidgilbertson you should have the option then. It could have possibly been that the backend was not receiving a heartbeat (how the W&B process lets the backend know the run is still alive) for some reason. If this is something you are regularly seeing would you mind creating a separate Issue and tagging me there so we can look into it more?

Thank you,
Nate

**Comment by botcs** [comment]
This seems to be a fairly essential and simple feature for elaborate early stopping policies.
Adding this method to the `Run` class results in the expected behaviour (so far)

Just insert after `.delete()`: https://github.com/wandb/wandb/blob/v0.19.9/wandb/apis/public/runs.py#L543-L571
```
@normalize_exceptions
def stop(self):
    """Stop a currently running run by sending a GraphQL mutation to update its state.

    This method stops the run by updating its state via a mutation request.
    If the run is not in the 'running' state, a warning is issued and no action is taken.
    After the mutation call, the run is reloaded to update its local state.
    """
    if self.state != "running":
        wandb.termwarn("The run is not currently running; stop operation aborted.")
        return

    mutation = gql(
        """
        mutation StopRun($id: ID!) {
            stopRun(input: {id: $id}) {
                clientMutationId
            }
        }
        """
    )
    self._exec(mutation, id=self.storage_id)
    # Reload the run to update the local state
    self.load(force=True)
```

**Comment by ArtsiomWB** [comment]
Hey @botcs, thank you for writing in. Since our SDK is open source you are more than welcome to submit a PR regarding this, and I will send it to our SDK team.
